src/Microsoft.ML.Transforms/TensorflowTransform.cs

```diff
@@ -0,0 +1,665 @@
+
+            _batchSize = args.BatchSize;
+            _session = LoadTFSession(args.ModelFile);
+            _bindings = new Bindings(args.Column, Source.Schema, this);
+        }
+
+        private TensorflowTransform(IHost host, ModelLoadContext ctx, IDataView input)
+           : base(host, input)
+        {
+            Host.AssertValue(ctx);
+
+            _batchSize = ctx.Reader.ReadInt32();
+
+            byte[] data = null;
+            ctx.TryLoadBinaryStream("TFModel", r =>
+            {
+                data = r.ReadByteArray();
+            });
+
+            var graph = new TFGraph();
+            graph.Import(data);
+            _session = new TFSession(graph);
+            _bindings = new Bindings(ctx, Source.Schema, this);
+        }
+
+        ~TensorflowTransform()
+        {
+            _session.CloseSession();
+            _session.DeleteSession();
+        }
+
+        public static TensorflowTransform Create(IHostEnvironment env, ModelLoadContext ctx, IDataView input)
+        {
+            Contracts.CheckValue(env, nameof(env));
+            var h = env.Register(RegistrationName);
+            h.CheckValue(ctx, nameof(ctx));
+            h.CheckValue(input, nameof(input));
+            ctx.CheckAtModel(GetVersionInfo());
+            return h.Apply("Loading Model", ch => new TensorflowTransform(h, ctx, input));
+        }
+
+        private TFSession LoadTFSession(string modelFile)
+        {
+            var graph = new TFGraph();
+            graph.Import(File.ReadAllBytes(modelFile), "");
+            return new TFSession(graph);
+        }
+
+        public override void Save(ModelSaveContext ctx)
+        {
+            Host.AssertValue(ctx);
+            ctx.CheckAtModel();
+            ctx.SetVersionInfo(GetVersionInfo());
+
+            ctx.Writer.Write(_batchSize);
+
+            var buffer = new TFBuffer();
+            _session.Graph.ToGraphDef(buffer);
+
+            ctx.SaveBinaryStream("TFModel", w =>
+            {
+                w.WriteByteArray(buffer.ToArray());
+            });
+            _bindings.Save(ctx);
+        }
+
+        public void DisposeTFSession()
+        {
+            _session.CloseSession();
+            _session.DeleteSession();
+        }
+
+        private ValueGetter<T> GetSrcGetter<T>(IRow input, int iinfo, int isrc)
+        {
+            return input.GetGetter<T>(_bindings.Infos[iinfo].SrcIndices[isrc]);
+        }
+
+        private ITensorValueGetter CreateTensorValueGetter<T>(IRow input, ColumnType type, int colIndex, TFShape tfShape)
+        {
+            if (type.IsVector)
+                return new TensorValueGetterVec<T>(input, colIndex, tfShape);
+            else
+                return new TensorValueGetter<T>(input, colIndex);
+        }
+
+        private ITensorValueGetter CreateTensorValueGetterVec(IRow input, TFDataType tfType, ColumnType type, int colIndex, TFShape tfShape)
```

Comment:
>CreateTensorValueGetterVec [](start = 35, length = 26)

Add a method similar to Tf2MlNetType() that returns a Type, and then you can reduce this method to to lines:

var type = Tf2SystemType(tfType);
return Utils.MarshalInvoke(CreateTensorValueGetter<int>, type, input, type, colIndex, tfShape);

Reply:
I changed what about the template type <int> in CreateTensorValueGetter<int>.

---
In reply to: [210656910](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210656910) [](ancestors = 210656910)